contracts/Launchpad.sol

```diff
@@ -0,0 +1,492 @@
+        launchTokenAddress[launchId] = tokenAddress;
+        launchTokenAddress[launchId + 1] = tokenAddress;
+    }
+
+    // VESTING
+
+    function enablePeriodicVesting(uint256 launchId, uint256 percentage) external onlyPhase1(launchId) isTokenAddressSet(launchId) restricted {
+        require(launchLinearVestingsDurations[launchId] == 0, "Linear vesting enabled");
+        require(getTotalUnlockedPercentage(launchId) + percentage <= 100, "Total percentage exceeded");
+        uint256 totalLaunchRaised = launchTotalRaised[launchId].add(launchTotalRaised[launchId + 1]);
+        uint256 contractBalance = IERC20(launchTokenAddress[launchId]).balanceOf(address(this));
+        uint256 decimals = ERC20(launchTokenAddress[launchId]).decimals();
+        if(decimals > 18) {
+            contractBalance = contractBalance.mul(10**uint(decimals - 18));
+        } else {
+            contractBalance = contractBalance.div(10**uint(18 - decimals));
+        }
+        require(contractBalance >=
+            totalLaunchRaised.mul(1e18).div(launchTokenPrice[launchId]).mul(getTotalUnlockedPercentage(launchId) + percentage).div(100),
+            "Not enough balance");
+
+        launchPeriodicVestingsPercentages[launchId].push(percentage);
+        emit PeriodicVestingEnabled(launchId, launchPeriodicVestingsPercentages[launchId].length - 1, block.timestamp);
+    }
+
+    function enableLinearVesting(uint256 launchId, uint256 duration) external onlyPhase1(launchId) isTokenAddressSet(launchId) restricted {
+        require(duration > 0);
+        require(launchPeriodicVestingsPercentages[launchId].length == 0, "Periodic vesting enabled");
+        uint256 totalLaunchRaised = launchTotalRaised[launchId].add(launchTotalRaised[launchId + 1]);
+        uint256 contractBalance = IERC20(launchTokenAddress[launchId]).balanceOf(address(this));
+        uint256 decimals = ERC20(launchTokenAddress[launchId]).decimals();
+        if(decimals > 18) {
+            contractBalance = contractBalance.mul(10**uint(decimals - 18));
+        } else {
+            contractBalance = contractBalance.div(10**uint(18 - decimals));
+        }
+        require(contractBalance >= totalLaunchRaised.mul(1e18).div(launchTokenPrice[launchId]), "Not enough balance");
+
+        launchLinearVestingsDurations[launchId] = duration;
+        launchLinearVestingsStartTimestamps[launchId] = block.timestamp;
+
+        emit LinearVestingEnabled(launchId, block.timestamp);
+    }
+
+    // WHITELISTING
+
+    function setEligibleUsersData(uint256 launchId, address[] calldata users, uint256[] calldata stakedAmounts) external onlyPhase1(launchId) restricted {
+        require(nextLaunchId >= launchId, "Wrong ID");
+        require(users.length == stakedAmounts.length, "Bad input");
+        require(launchStartTime[launchId] > block.timestamp, "Launch started");
+
+        uint totalWeight = 0;
```

Comment:
This needs to be assigned with `launchFundsToRaise[launchId].div(launchBaseAllocation[launchId])` to allow for calling this function in batches. I think thousands of entries per array in arguments can hit the gas limits.